pkgs/by-name/pr/proton-ge-bin/package.nix

```diff
@@ -0,0 +1,50 @@
+{ lib
+, stdenvNoCC
+, fetchurl
+, writeScript
+}:
+stdenvNoCC.mkDerivation (finalAttrs: {
+  pname = "proton-ge-bin";
+  version = "GE-Proton9-1";
+
+  src = fetchurl {
+    url = "https://github.com/GloriousEggroll/proton-ge-custom/releases/download/${finalAttrs.version}/${finalAttrs.version}.tar.gz";
+    hash = "sha256-wCIffeayOy3kEwmIKB7e+NrliuSpKXoVYC334fxVB3U=";
+  };
+
+  buildCommand = ''
+    runHook preBuild
+    mkdir -p $out/{bin,opt}
+    tar -C $out/opt --strip=1 -x -f $src
+    ln -s $out/opt/toolmanifest.vdf $out/bin/toolmanifest.vdf
+    install -Dm644 $out/opt/compatibilitytool.vdf $out/bin/compatibilitytool.vdf
+    substituteInPlace $out/bin/compatibilitytool.vdf \
+      --replace-fail '"install_path" "."' '"install_path" "${placeholder "out"}/opt"'
```

Comment:
(Continuation of https://github.com/NixOS/nixpkgs/pull/294532#discussion_r1523656697)

> When not using `makeBinPath` we either can drop the whole package content at `$out`, or drop it at `$out/{something}` like suggested. It would be just needed then that `compatibilitytool.vdf` and `toolmanifest.vdf` are set correctly.

That'd seem like the most sensible to me. In a separate output of course.

> So by using `makeBinPath` we need at least
> 
> ```
> result/
> └── bin
>      └── compatibilitytool.vdf
> ```

If makeBinPath is the only reason for doing this, then let's do away with it. These are not binary executables, so they have no business being in `bin/`.

> I am personally up here for any change, dropping everything just in `$out` might be the easiest but was kind of declined on the other PR by [#189398 (comment)](https://github.com/NixOS/nixpkgs/pull/189398#discussion_r962180579)

I'm not entirely sure what they're trying to say there (cc @firestack) but I think they mean that dumping the package into the output would result in a "package" that doesn't conform to our standards where each package internally mimics FHS-isms (i.e. `bin/` `share/`) but rather dumps everything at the top-level. Including it in a mkEnv would likely produce weird results.

I share that concern and consider it a blocker. We will not have top-level pkgs with weird directory structures.

I see two ways of addressing it:

1. Use the `out` output but place the compat tool's contents in `share/steam/compatibilitytools.d/$name/`. This conforms to the FHS-isms and is compatible with mkEnv and the like. We could even have a global compatibility tool dir by doing this but I fail to see the use-case for that as it only concerns Steam.
2. Dump the contents at the top-level of a separate `compatibilitytool` output. We could symlink that to `$out/share/steam/compatibilitytools.d/$name/` too now that I think about it.

I am partial to 2. because of its simplicity, idiomacy and flexibility. It also allows for sanity checks in extraCompatPackages as we could define that all drvs passed to it must have a `compatibilitytool` output. I wouldn't be opposed to 1. but at this point I think 2. has quite a few more merits.

Reply:
What do you think of option 2.?

Reply:
Not very sure. What's the point of separating outputs?

Reply:
This would simplify packaging a lot (bin packages could basically just be unpacked into the output; no patching required) and provide a sane standard that can be checked for at eval time.

Only code that explicitly looks for the `compatibilitytools` output (like the extraCompatTools NixOS option would) would be able to discover these files and it'd then obviously be their responsibility to handle them appropriately.

When a user would accidentally install `proton-ge-bin`, into some env, the `out` (or `bin`) outputs are used. If we were to put the compatibility tool files into a separate output that isn't either of those, including the package inside an env would be a no-op; no global paths littered with files.

Reply:
Are you talking about revolutionizing packaging (and starting with a totally irrelevant package) or are you talking about improving this package specifically?

There is no case where a user would have to add proton-ge to an environment (e.g. `environment.systemPackages`), even less so an FHS environment. Steam (or any other tool) cannot pick them up from the environment automagically, and I feel like we're slowly entering bikeshedding territory for no reason.

The `extraCompatModules` option (https://github.com/NixOS/nixpkgs/blob/nixos-unstable/nixos/modules/programs/steam.nix#L47C14-L47C82)

```nix
STEAM_EXTRA_COMPAT_TOOLS_PATHS = makeBinPath cfg.extraCompatPackages;
```

would return $out/bin for proton-ge, and unless the module itself has changed, I fail to see a reason to provide a $out/bin output and a $out/whatever output. That said, I will not provide any further input on this matter. If you have a choice of default paths (that still provides an $out/bin), lets go with that. Otherwise I'd like to change $out/opt and get this merged.

Reply:
> Dump the contents at the top-level of a separate compatibilitytool output.

@Atemu is there a package that you have in mind, that does something like that? I have correctly a bit problem to wrap my head around that (tho I guess understand what you try to archive)

Reply:
> Are you talking about revolutionizing packaging (and starting with a totally irrelevant package) or are you talking about improving this package specifically?

I'm talking about setting a good standard for Steam compatibilitytools in Nixpkgs.

> There is no case where a user would have to add proton-ge to an environment (e.g. `environment.systemPackages`), even less so an FHS environment.

Believe me, people will attempt precisely that and be confused why it doesn't work.

> The `extraCompatModules` option (https://github.com/NixOS/nixpkgs/blob/nixos-unstable/nixos/modules/programs/steam.nix#L47C14-L47C82)
> 
> ```nix
> STEAM_EXTRA_COMPAT_TOOLS_PATHS = makeBinPath cfg.extraCompatPackages;
> ```

As @Shawn8901 mentioned, changing `extraCompatPackages` is on the table. That option has exactly 0 packages it could be used with in Nixpkgs atm. This would be the first. Changing it to fit the needs of the packages to be used in it now that we know about the reality of their packaging a little more is absolutely something I'd expect to do.

I was very busy the past few weeks/months and I didn't catch that/think about it too much and the PR was merged without me having the chance to give proper input on that topic or else I would have discussed it there.

It's not a big deal; we can still easily change it now.

> If you have a choice of default paths (that still provides an $out/bin), lets go with that. Otherwise I'd like to change `$out/opt` `$opt/share` and get this merged.

That's effectively option 1. which I'm fine with. The current form I'm not fine with; we won't be putting things other than binary executables into `bin/`.

> @Atemu is there a package that you have in mind, that does something like that? I have correctly a bit problem to wrap my head around that (tho I guess understand what you try to archive).
> I could try to prepare it then.

A precedent would be mesa's `drivers` output. Mesa provides a few non-driver libraries such as libgbm and those are in `out`. Drivers for `hardware.opengl.packages` are taken explicitly from `mesa.drivers`.

> Is assume that our fhs package wrapping is kind of doing the same that you explain (do having a base package and then around some thing that builds up the context (in our case here the compat tooling) what also is cool then, that other compat derivations could use the same code when added to build their valid compat definition

Sorry, could you reword that? I'm not sure I understand what you're trying to say but I don't think it's what I'm saying. I'll draft up a PR based on this one to show what I mean.

Reply:
I've based https://github.com/NixOS/nixpkgs/pull/296009 on your PR. Haven't tested yet as I can't use my desktop PC until tomorrow.

Reply:
> > There is no case where a user would have to add proton-ge to an environment (e.g. `environment.systemPackages`), even less so an FHS environment.
> 
> Believe me, people will attempt precisely that and be confused why it doesn't work.

And it shouldn't work. proton-ge is a compatibility tool, not an executable.

Also, what about tools like steamtinkerlaunch that [also go in compatibilitytools.d](https://github.com/sonic2kk/steamtinkerlaunch/wiki/Steam-Compatibility-Tool)?

Reply:
As soon as [I can get it to work](https://github.com/NixOS/nixpkgs/issues/210018), STL will be adjusted too :)

Reply:
> Sorry, could you reword that? I'm not sure I understand what you're trying to say but I don't think it's what I'm saying. I'll draft up a PR based on this one to show what I mean.

By seeing the pr from you i, i guess i wasnt talking the same, i did know that there are different outputs in a derivation, but never had digged into that in more detail. 
So i will take a look on the pr and report back :)